### PR TITLE
add frontend e2e test for create route

### LINKED ISSUE #975: test: add frontend e2e test for create route
use the current test framework to write web E2E tests:

- [ ] create a route according to the default steps

       1. input route `Name`
       2. input route `Path`
       next
       3. input upstream `Node ip`
       next
       4. click submit
       5. confirm page return

- [ ] verify the route
       check on the `/routes/list` page
- [ ] delete the route

### PR DESCRIPTION
**Pull request by idbeta**
fixed #975

Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Frontend E2E test
- [ ] New feature provided
- [ ] Improve performance

### CODE REVIEW

**Comment by moonming** [comment]
why e2e test cases in “src” dir？

File: web/src/e2e/Route.e2e.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  routelist:'.ant-layout-sider-children ul li:nth-child(2)',
+  button_createroute:'.ant-pro-table-toolbar-option button',
+  button_next:'.ant-row-end .ant-col:nth-child(2)'
+};
+
+describe('Route', () => {
+
+  beforeAll(async () => {
+    browser = await puppeteer.launch({
+      headless: false,
```
Comment (moonming) [inline web/src/e2e/Route.e2e.js:42]: why not headless？

Comment (idbeta) [reply web/src/e2e/Route.e2e.js]: fixed

**Comment by idbeta** [comment]
> why e2e test cases in “src” dir？

I wrote the use case based on this [document](https://github.com/apache/apisix-dashboard/blob/master/web/src/e2e/README.md), it seems a little strange. @LiteSun please answer this when you are free

**Comment by membphis** [comment]
> > why e2e test cases in “src” dir？

this is unacceptable. we need to change it

**Comment by juzhiyuan** [comment]
> why e2e test cases in “src” dir？

https://github.com/ant-design/ant-design-pro/tree/master/src/e2e

The frontend's structure is according to Ant Design Pro.

cc @membphis @moonming

**Comment by membphis** [comment]
image

File: .github/workflows/frontend-e2e-test.yml [hunk 2]
```diff
@@ -31,9 +41,20 @@ jobs:
...
           node-version: 14.x
 
-      - name: Install dependencies
+      - name: Setup golang environment
+        uses: actions/setup-go@v1
+        with:
+          go-version: '1.13'
+
+      - name: Start manager-api
+        working-directory: ./
```
Comment (membphis) [inline .github/workflows/frontend-e2e-test.yml:50]: how about this way?

```
        working-directory: ./api
        run: nohup go run . &

File: web/src/e2e/Route.e2e.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    // access route list page
+    await page.waitForSelector(domSelectors.route_list)
+    await page.click(domSelectors.route_list)
+    await page.content();
+    // create route
+    await page.waitForSelector(domSelectors.button_createroute)
+    await page.click(domSelectors.button_createroute)
+    await page.content();
+
+    await page.waitForSelector(domSelectors.input_routename);
```
Comment (membphis) [inline web/src/e2e/Route.e2e.js:60]: need a comment

File: web/src/e2e/Route.e2e.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    await page.type(domSelectors.input_routename, "test_route_by_ui_autotest");
+    await page.focus(domSelectors.input_routepath);
+    await page.keyboard.press( 'Backspace' );
+    await page.keyboard.press( 'Backspace' );
+    await page.type(domSelectors.input_routepath, "/testpath_by_ui_autotest");
+    await page.click(domSelectors.button_next);
+
+    await page.waitForSelector(domSelectors.input_node_host);
```
Comment (membphis) [inline web/src/e2e/Route.e2e.js:68]: ditto

**Review by membphis — CHANGES_REQUESTED** [review]
ready for review?

**Comment by idbeta** [comment]
> ready for review?

I want to try the `delete route` action again

File: .github/workflows/frontend-e2e-test.yml (hunk 2 above)
Comment (idbeta) [reply .github/workflows/frontend-e2e-test.yml]: agree, fixed.

**Comment by membphis** [comment]
welcome review @LiteSun @moonming @liuxiran

File: web/src/e2e/Route.e2e.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+/* eslint-disable import/no-extraneous-dependencies */
+const puppeteer = require('puppeteer');
+
+const {
+  setupLogin
+} = require('./service')
+
+let browser;
+const domSelectors = {
+  input_routename: '#name',
```
Comment (LiteSun) [inline web/src/e2e/Route.e2e.js:27]:
```suggestion
  inputRouteName: '#name',
```

please use CamelCase to keep same style.

**Comment by LiteSun** [comment]
One PR one thing. 

 I think the PR needs to be split into two PRs. 
1. add create route E2E test
2. modify the YAML file to start the local manager-api for test .

What do you think? @membphis @juzhiyuan @idbeta

File: web/src/e2e/Route.e2e.js (hunk 3 above)
Comment (idbeta) [reply web/src/e2e/Route.e2e.js]: fixed.

File: web/src/e2e/Route.e2e.js (hunk 4 above)
Comment (idbeta) [reply web/src/e2e/Route.e2e.js]: fixed.

File: web/src/e2e/Route.e2e.js (hunk 5 above)
Comment (idbeta) [reply web/src/e2e/Route.e2e.js]: agree, fixed.

File: web/src/e2e/Route.e2e.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  inputNodehost: '#nodes_0_host',
+  inputNodeport: '#nodes_0_port',
+  inputNodeweight: '#nodes_0_weight',
+  routeList: '.ant-layout-sider-children ul li:nth-child(2)',
+  buttonCreateroute: '.ant-pro-table-toolbar-option button',
+  buttonNext: '.ant-row-end .ant-col:nth-child(2)',
+  successIco: '.ant-result-success .anticon-check-circle',
+  buttonReturnRoutelist: '.ant-result-extra .ant-btn-primary'
+};
+
```
Comment (moonming) [inline web/src/e2e/Route.e2e.js:38]: For me these are too difficult to maintain

Comment (moonming) [reply web/src/e2e/Route.e2e.js:38]: we need to find a better way.

**Comment by idbeta** [comment]
There is a new frontend E2E test technology solution that needs to be evaluated. After the evaluation, I will decide whether to open this PR.

File: web/src/e2e/Route.e2e.js (hunk 6 above)
Comment (moonming) [reply web/src/e2e/Route.e2e.js:38]: please take a look at https://github.com/getgauge/taiko

Comment (LiteSun) [reply web/src/e2e/Route.e2e.js:38]: > please take a look at https://github.com/getgauge/taiko

ok, will try to find a better way. refer issue: https://github.com/apache/apisix-dashboard/issues/1036

**Pull request closed without merging**